### ISSUE TITLE
sysroutes_dnet_find_interfaces strncpy of sockstruct_addr structures tripping GCC stringop-truncation warnings

### ISSUE THREAD

**Issue by berney** [nmap/nmap#2186]
**Describe the bug**

The function `sysroutes_dnet_find_interfaces`, in `libnetutil/netutil.cc`, does `strncpy` of the str char array returned from `inet_ntop_ez`, to buffer of equal length, which `strncpy` doesn't guarantee to null terminate, causing GCC stringop-truncation warning. There are two instances of this.

`INET6_ADDRLEN` = 46, which is taking into account null terminating largest string presentation of an IPv6 address, so the `strncpy` should always be null-terminated. GCC's stringop-truncation check can't know this though.

```
Compiling libnetutil
cd libnetutil && make
make[1]: Entering directory '/var/tmp/portage/net-analyzer/nmap-9999/work/nmap-9999/libnetutil'
x86_64-gentoo-linux-musl-g++ -c -I/usr/include -I../liblua -I../libdnet-stripped/include -I../libssh2/include  -I../nbase -I../nsock/include -DHAVE_CONFIG_H  -mtune=generic -O2 -pipe -static -static-libstdc++ -static-libgcc -Wall -fno-strict-aliasing netutil.cc -o netutil.o

... <snip> ...

In function 'dnet_collector_route_nfo* sysroutes_dnet_find_interfaces(dnet_collector_route_nfo*)',
    inlined from 'sys_route* getsysroutes_dnet(int*, char*, size_t)' at netutil.cc:1678:37,
    inlined from 'sys_route* getsysroutes(int*, char*, size_t)' at netutil.cc:1707:29:
netutil.cc:1612:14: warning: 'char* strncpy(char*, const char*, size_t)' specified bound 46 equals destination size [-Wstringop-truncation]
 1612 |       strncpy(destbuf, inet_ntop_ez(&dcrn->routes[i].dest, sizeof(dcrn->routes[i].dest)), sizeof(destbuf));
      |       ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
netutil.cc:1613:14: warning: 'char* strncpy(char*, const char*, size_t)' specified bound 46 equals destination size [-Wstringop-truncation]
 1613 |       strncpy(gwbuf, inet_ntop_ez(&dcrn->routes[i].gw, sizeof(dcrn->routes[i].gw)), sizeof(gwbuf));
      |       ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**To Reproduce**
Steps to reproduce the behavior, including command-line options.

```
cd libnetutil
x86_64-gentoo-linux-musl-g++ -c -I/usr/include -I../liblua -I../libdnet-stripped/include -I../libssh2/include  -I../nbase -I../nsock/include -DHAVE_CONFIG_H  -mtune=generic -O2 -pipe -static -static-libstdc++ -static-libgcc -Wall -fno-strict-aliasing netutil.cc -o netutil.o
```

**Expected behavior**
No string truncation compiler warnings.

**Version info (please complete the following information):**
 - OS: Gentoo Linux 

Portage 3.0.8 (python 3.7.9-final-0, default/linux/amd64/17.0/musl/hardened, gcc-9.3.0, musl-1.2.1, 4.19.128-microsoft-standard x86_64)
With x86_64-gentoo-linux-musl-g++ (Gentoo Hardened 9.3.0-r1 p3) 9.3.0
sys-libs/musl:            1.2.1::gentoo

 - Output of `nmap --version`:

network-analyzer/nmap-9999
nmap git commit 169d7e5a9

 - Output of `nmap --iflist`
N/A

**Additional context**

**Comment by nnposter** [comment]
Both buffers, `destbuf` and `gwbuf`, are of size `INET6_ADDRSTRLEN`, which happens to be also the size of the buffer returned by `inet_ntop_ez()`, so if the buffers were to logically overflow, it would happen there first. However, this will not happen because the buffer in `inet_ntop_ez()` is populated by standard `inet_ntop()`, which is never going to generate a string longer than `INET6_ADDRSTRLEN` (counting the terminating `\0`). See https://pubs.opengroup.org/onlinepubs/009695399/functions/inet_ntop.html

In summary, the use of `strncpy()` is likely a form of defensive coding. Using straight `strcpy()` would be perfectly safe here as well.

**Comment by berney** [comment]
Yeah, the `strncpy()` should always be safe, but it would be nice to silence the warning.

**Comment by nnposter** [comment]
I am using GCC 7.5.0, which does not complain. Can you test the following patch?
```diff
--- a/libnetutil/netutil.cc	2020-01-14 17:33:58.000000000 -0700
+++ b/libnetutil/netutil.cc	2020-11-24 20:02:15.142409200 -0700
@@ -1677,8 +1677,8 @@
       char destbuf[INET6_ADDRSTRLEN];
       char gwbuf[INET6_ADDRSTRLEN];
 
-      strncpy(destbuf, inet_ntop_ez(&dcrn->routes[i].dest, sizeof(dcrn->routes[i].dest)), sizeof(destbuf));
-      strncpy(gwbuf, inet_ntop_ez(&dcrn->routes[i].gw, sizeof(dcrn->routes[i].gw)), sizeof(gwbuf));
+      Strncpy(destbuf, inet_ntop_ez(&dcrn->routes[i].dest, sizeof(dcrn->routes[i].dest)), sizeof(destbuf));
+      Strncpy(gwbuf, inet_ntop_ez(&dcrn->routes[i].gw, sizeof(dcrn->routes[i].gw)), sizeof(gwbuf));
       /*
       netutil_error("WARNING: Unable to find appropriate interface for system route to %s/%u gw %s",
         destbuf, dcrn->routes[i].netmask_bits, gwbuf);
```

**Comment by berney** [comment]
I tested your patch and it works, no compiler warnings in same setup as above (gcc-9.3.0). `Strncpy` always doing the null termination, seems to make the GCC stringop-truncation heuristics happy.

Locally, I had also tried this patch, which works, but I prefer yours.

```diff
diff --git a/libnetutil/netutil.cc b/libnetutil/netutil.cc
index 84905dd84..107d4e90e 100644
--- a/libnetutil/netutil.cc
+++ b/libnetutil/netutil.cc
@@ -1609,8 +1609,11 @@ static struct dnet_collector_route_nfo *sysroutes_dnet_find_interfaces(struct dn
       char destbuf[INET6_ADDRSTRLEN];
       char gwbuf[INET6_ADDRSTRLEN];

+      #pragma GCC diagnostic push
+      #pragma GCC diagnostic ignored "-Wstringop-truncation"
       strncpy(destbuf, inet_ntop_ez(&dcrn->routes[i].dest, sizeof(dcrn->routes[i].dest)), sizeof(destbuf));
       strncpy(gwbuf, inet_ntop_ez(&dcrn->routes[i].gw, sizeof(dcrn->routes[i].gw)), sizeof(gwbuf));
+      #pragma GCC diagnostic pop
       /*
       netutil_error("WARNING: Unable to find appropriate interface for system route to %s/%u gw %s",
         destbuf, dcrn->routes[i].netmask_bits, gwbuf);
```

I haven't tested other GCC versions, but I think earlier versions might warn if they don't understand the pragma, which defeats the purpose.

**Comment by nnposter** [comment]
Thank you for validating. The "Strncpy" patch has been committed as r38147.

**Comment by goarblt** [comment]
hello
https://www.evil.com/
